mesonbuild/arguments.py

```diff
@@ -0,0 +1,126 @@
+# SPDX-License-Identifier: Apache-2.0
+# Copyright © 2025 Intel Corporation
+
+"""An abstract IR for compile arguments.
+
+This provides us with a way to pass arguments around in a non-compiler specific
+way internally, and lower them into a non-abstract format when writing them in
+the backend.
+"""
+
+from __future__ import annotations
+import dataclasses
+import typing as T
+
+if T.TYPE_CHECKING:
+    from typing_extensions import TypeAlias
+
+    # A Union of all Argument types
+    Argument: TypeAlias = T.Union[
+        'Opaque', 'Warning', 'Error', 'Define', 'Undefine', 'LinkerSearch',
+        'LinkLibrary', 'Rpath',
+    ]
+
+
+@dataclasses.dataclass
+class Opaque:
+
+    """An opaque argument.
+
+    This is an argument of unknown type, and Meson will do nothing but proxy it
+    through, except to apply a prefix to the value if it thinks it's necessary.
+
+    :param value:
+    """
+
+    value: str
+
+
+@dataclasses.dataclass
+class Warning:
+
+    """A compiler warning.
+
+    :param target: The warning to enable or disable. This will be stored as
+        given (ie, we don't try to convert between compilers).
+    :param enable: If true then enable the warning, otherwise suppress it.
+    """
+
+    target: str
+    enable: bool
+
+
+@dataclasses.dataclass
+class Error:
+
+    """A compiler error.
+
+    :param target: The warning to enable or disable. This will be stored as
+        given (ie, we don't try to convert between compilers).
+    """
+
+    target: str
+
+
+@dataclasses.dataclass
+class Define:
+
+    """A pre-processor define.
+
+    :param target: The value to define.
+    :param value: An optional value to set the define to. If undefined them the
+        value will be defined with no value.
+    """
+
+    target: str
+    value: T.Optional[str]
+
+
+@dataclasses.dataclass
+class Undefine:
```

Comment:
For warnings we have enable=True, why not the same for Define?

Reply:
In fact since `-Dfoo` and `-Dfoo=` are the same, could `Undefine('foo')` be replaced by `Define('foo', value=None)`?

Reply:
You're right, sorry. I meant that `-Dfoo` and `-Dfoo=1` could be canonicalized to a single form (`Define('foo', '1')` in the abstract form, whereas the concrete form could be either) and `value=None` could be used for undefining.

Reply:
Yes... OTOH  you do want to collapse `-Dfoo -Dfoo=1` to just one of them.  Setting `value: `1`` would also make `__eq__` easier to implement.

Maybe some kind of `have_equals: bool` could be added to make round-trip conversion possible, and it would be ignored by `__eq__`.

Reply:
I don't know that we have to guarantee we pass *exactly* the same arguments as long as we pass equivalent arguments. Especially since we already attempt to do some level of de-duplication. I'm perfectly fine with `-Dfoo=` becoming `-Dfoo=1`.